### PR TITLE
Ct 278 roles and permissions

### PR DESCRIPTION
**Pull request by EdwardAndress**
For CT-278

Implement permissions using either Pundit, CanCan, or another gem.

- I chose to do a spike with Pundit

https://github.com/elabs/pundit

- To this end, I created a simple feature test, where only assigners are able to view the New Case form.
- This works quite nicely and we can add custom flash messages using locales
- Pundit also comes with it's own rspec matchers as you will see!
- In addition, we will be able to use Pundit to return specific scopes on the index action, should we decide to show different users different list views, based on their role (although I suspect we are likely to go somewhat beyond that).
- There is also a method to ensure authorisation is checked after each action, should we decide to use it.

### CODE REVIEW

**Comment by EdwardAndress** [comment]
If Pundit provides all the functionality we need and the syntax is not objectionable I vote we go with it @misaka and @aliuk2012

**Comment by aliuk2012** [comment]
I suppose you could also test to ensure that Assigners cannot access the assignments edit page

**Comment by EdwardAndress** [comment]
@aliuk2012 - sure, we could do that and I am happy to stick it in if we feel it's appropriate.  The idea, though, was to just add the simplest possible example of a 'policy' at this stage.

The policy for that edit page might be that you can only see the edit page for assignments where you are the assignee...

**Comment by aliuk2012** [comment]
Good point about it would be restricted to drafters and the team its assigned for.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,12 +1,26 @@
...
   def set_user
     @user = current_user
   end
 
+  private
+
+  def user_not_authorized(exception)
+    policy_name = exception.policy.class.to_s.underscore
+
+    flash[:error] = t "pundit.#{policy_name}.#{exception.query}"
```
Comment (aliuk2012) [inline app/controllers/application_controller.rb]: You need to replace ':error' with ':alert' to get the error message styled properly

Comment (EdwardAndress) [inline app/controllers/application_controller.rb]: Done!

**Comment by aliuk2012** [comment]
😠  Flicker again - Subject text from factory girl longer than 80chars

https://travis-ci.org/ministryofjustice/correspondence_tool_staff/builds/195920336

File: app/views/cases/index.html.slim [hunk 2]
```diff
@@ -1,8 +1,9 @@
 = render partial: 'layouts/header', locals: {page_heading: t('.page_heading')}
 
-.grid-row
-  .column-full
-    = link_to t('common.new_case_button'), new_case_path, class: 'button'
+- if policy(:case).new?
```
Comment (misaka) [inline app/views/cases/index.html.slim]: As discussed in person, I think that `new?` doesn't communicate very well what we're checking for here, `create_new?` or something similar would make this a little bit clearer.

File: app/policies/case_policy.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class CasePolicy
+
+  attr_reader :user, :case
+
+  def initialize(user, kase)
+    @user = user
+    @case = kase
+  end
+
+  def new?
```
Comment (misaka) [inline app/policies/case_policy.rb]: Shouldn't we have a `create?` here too? Should these be separate methods? Paired with the comment that `new?` in the view could be better communicated, one idea is to have `create_new?` in the policy, and then to alias `new?` and `create?` to it, to DRY things up.

**Comment by EdwardAndress** [comment]
@misaka I have made some changes as per the chat we had earlier.

@aliuk2012 you may want to check out the second commit here

**Pull request merged**